core/src/main/scala/kafka/cluster/Partition.scala

```diff
@@ -235,10 +233,20 @@ class Partition(val topic: String,
   def updateReplicaLogReadResult(replicaId: Int, logReadResult: LogReadResult) {
     getReplica(replicaId) match {
       case Some(replica) =>
+        // No need to calculate low watermark if there is no delayed DeleteRecordsRequest
+        val oldLeaderLW = if (replicaManager.delayedDeleteRecordsPurgatory.delayed() > 0) lowWatermarkIfLeader else -1L
         replica.updateLogReadResult(logReadResult)
+        val newLeaderLW = if (replicaManager.delayedDeleteRecordsPurgatory.delayed() > 0) lowWatermarkIfLeader else -1L
```

Comment:
Thinking a bit more. It seems that a delayedDeleteRecord could be added right after the if check. In the next updateReplicaLogReadResult() call, LW may not advance any more. Then the response will be delayed. Perhaps, if (replicaManager.delayedDeleteRecordsPurgatory.delayed() > 0) , we could just always call tryCompleteDelayedRequests()?